readthedocs/builds/models.py

```diff
@@ -1359,51 +1357,99 @@ def move(self, steps):
         total = self.project.automation_rules.count()
         current_priority = self.priority
         new_priority = (current_priority + steps) % total
+        self.priority = new_priority
+        self.save()
 
-        if current_priority == new_priority:
-            return False
+    def _change_priority(self):
```

Comment:
I think the logic of this method is unnecessary complex and can be re-written in a simpler way.

1. if it's a new rule added: all the `priority` will be increased in 1 and the new rule will have `priority=0`
2. if we are moving a rule up/down we need to update the priority of two rules only, by swapping them: the one we are moving and the upper/down one depending if we move it up or down.

Consider this example for case 2). If you have 5 rules:

0
1
2
3
4

and you want to move the 2 down, you just need to update the rule 2 with +1 and the rule 3 with -1. In pseudo-code:

```python
current_priority = 2
move_up = -1
move_down = 1

rule_2 = self.project.automation_rules.get(priority=current_priority).update(priority=current_priority + move_down)
rule_3 = self.project.automation_rules.get(priority=current_priority + move_down).update(priority=current_priority)
```

Reply:
I plan to reuse this logic for redirects, there we won't have these two cases, an item (redirect) can be inserted at any place, or can be created at any place (this will ease the update operations in the API). Our UI can also benefit of this, we can implement drag and drop operations instead of having to manually move up and down the rule several times.

Reply:
The "new" logic is the same that we had before, with the extra case that a new rule can be inserted at any position.

Reply:
I understand. My proposal can be easily expanded to support those cases you are mentioning in a pretty simple way. Basically:

* inserting/creating a new item at any position would just get all the items with higher priority than the position created and increase the priority in 1
```python
# Verify the position is correct
assert 0 <= inserted_at <= self.project.automation_rules.count() + 1

# Update items
self.project.automation_rules.filter(priority__gte=inserted_at).update(priority=F("priority") + 1)

# Insert new item
self.project.automation_rules.create(priority=inserted_at)
```

* move an item to any other position: swap items priority from those positions between them 
```python
# Example: move item 2 to 4
# Swap priorities between items
current_position = 2
ending_position = 4

# Verify the positions are correct or return 400
assert 0 <= curent_position <= self.project.automation_rules.count()
assert 0 <= ending_position <= self.project.automation_rules.count()

# Get current and update priority with ending
rule_current = self.project.automation_rules.get(priority=current_position)
rule_current.priority = ending_position

# Get ending and update priority with current
rule_ending = self.project.automation_rules.get(priority=ending_position)
rule_ending.priority = current_position

rule_current.save()
rule_ending.save()
```

This is my suggestion to avoid the unnecessary complexity of this method. What do you think? Do you see any problem with this approach? Would you mind giving it a try? You have already created the test cases, so it shouldn't be too hard.

Reply:
I think this logic could be simplified a bit with something like what @humitos is describing here, it feels we should be able to do this mostly with bulk operations.

I'll throw another wrench into the plan though: giving the users the ability to manually edit the priority number will be really useful and maybe _required_ for redirects.

That is, if I have hundreds of redirects and for some reason need to add a new redirect squarely in the middle of the list, I would have to click hundreds of arrows to re-sort each rule I add to the list.

Especially for redirects, where users might want completely arbitrary priorities, we might have to get more creative with this pattern. In fact, this might disallow us from altering the priorities at all (at least the ones the users set), as the maintainer might have assigned a specific meaning to a priority number (redirects for pages from our old docs start at priority 5000)

Reply:
@humitos there is no need for swapping at the moment, what we want to have support for is insertion. The first part of the algorithm that you described (inserting a new item at any position) is basically what the current code does, there is some additional logic in order to avoid the unique constraints of SQLite.

@agjohnson bulk operations is what the current algorithm does more or less, but as mentioned in my other above, there is a restriction with SQLite.


https://github.com/readthedocs/readthedocs.org/blob/e8ad153cba9f058acd297567857cd9746c06dfaf/readthedocs/builds/models.py#L1441-L1445

Allowing to set the priority explicitly is what this algorithm provides, which is insertion at any position.

Reply:
:+1: I don't have many opinions on the algorithm. It is a little hard to follow to me, but I also don't have anything different to suggest.

I couldn't find the Cloudflare option for changing priority (other than dragging), but finally found it under the option menu. It prompts to enter the new item position number:

![image](https://github.com/readthedocs/readthedocs.org/assets/1140183/391814dc-a0bd-46b9-96ab-b5eaef4cbf6f)

This is good enough, though arbitrary values seem like the best UX overall (though also the hardest to accomplish for us).

Reply:
> It is a little hard to follow to me

This was the main reason why I suggested a simplified approach 😄